pkg/resource/deploy/source_eval.go

```diff
@@ -349,6 +503,15 @@ func (rm *resmon) RegisterResource(ctx context.Context,
 	parent := resource.URN(req.GetParent())
 	protect := req.GetProtect()
 
+	provider := req.GetProvider()
+	if custom && !providers.IsProviderType(t) && provider == "" {
+		ref, err := rm.defaultProviders.getDefaultProviderRef(t.Package())
```

Comment:
There's a lot of potential for deadlock here. In general there should be no blocking in the resource monitor that can't be canceled by closing `rm.cancel`. gRPC is going to wait for all in-flight RPCs to drain before shutting down the server and the `defaultProviders` server is going to exit if it gets canceled mid-operation.

We should either be returning a channel here that we `select` over `with <-rm.cancel` like we do on line 539, so we can report that the resource monitor is shutting down, or we should return some sort of well-known error that indicates that the resmon is shutting down.